Improved Kinematics

liamgilligan opened the pull request:
### This PR changes the Kinematics node in four major ways:

# 1. Improved Kinematics methodology
This PR changes the process by which the robot determines how much thrust each motor produces to a mathematically and physically sound method. It does this by encoding the resultant thrust and torque each motor will produce into a matrix and generating its [Moore-Penrose](https://en.wikipedia.org/wiki/Moore%E2%80%93Penrose_inverse) (pseudo)inverse. When this inverse matrix is multiplied by the pilot's desired twist vector, we are given a list of thrusts in newtons that each motor should produce in order to achieve the multiplied twist vector.

Moore-Penrose inverses have multiple important properties in respect to the problem of kinematics:

1. It finds the minimum Euclidean norm solution. In the case of kinematics, this means that it will produce the most efficient set of thrusts in order to achieve the desired twist; all other sets of thrusts will necessarily have more cancellation of force and torque, resulting in less efficiency.
2. It is a matrix based solution, which means it can easily and efficiently be [GPU accelerated](https://cupy.dev/) on the deck.

# 2. Variable Center of Mass
While the Moore-Penrose inverse method described above allows for precise control of the robot, it's main limitation is it requires that the position of the Center of Mass be accurate. One could imagine that as the Center of Mass moves away from its expected location (in practice due to lifting a heavy object) the actual torque produced by each motor will stray further and further from those prescribed by the matrix math (due to a changing radius).

In order to remedy this, the ability for a moving Center of Mass using ROS parameters was added. While this could potentially be controlled automatically in the future, the intended use was for the pilot to be able to [trim](https://en.wikipedia.org/wiki/Car_model#Trim_level) the Center of Mass of the robot during flight.

While the kinematics allows for arbitrary Center of Mass shifts, it is probably most advantageous to only give the pilot the ability to shift the center of mass along the line connecting the normal Center of Mass position to the Claw, as that is line the Center of Mass will actually shift along when lifting an object (a CoM reset button would also prove useful). Regardless of the implementation, the kinematics allows for it, so this can be tweaked in the future.

# 3. Thrust Limiting
The [Blue Robotics T200](https://bluerobotics.com/store/thrusters/t100-t200-thrusters/t200-thruster-r2-rp/) have several limitations, one of which being their thrust limits.

The motors have separate forward and reverse thrust limit, and we must take that into account. The methodology is to find a scaling factor for each motor that would scale it to its maximum thrust in the direction it is going. We then select the minimum of those possible scalars, as any of the other scaling factors will scale some of the other motors past their maximums.

# 4. Current Limiting
The robot has two main current limits which must be taken into account:

1. Total current draw must not exceed 70A
2. The current draw on each ESC must not exceed 40A

To do this we employ an identical method as we do for thrust limiting. We calculate the maximum scaling factors such that all the motors together draw 70A, the motors on the first ESC draw 40A, and the motors on the second ESC also draw 40A. We then select the minimum of those scaling factors.

Of course, to calculate any of those scaling factors is slightly more complicated. In the case of the thrust limiting, it is very easy as thrust is proportional to thrust (the proof of this is left to the reader as an exercise). However, there is a nonlinear relationship between current draw and thrust. We model that relationship with a 5th degree polynomial.

So for a given list of motor thrusts, we can do some algebraic manipulation to receive another 5th degree polynomial function, except this time it is a function of the desired scaling factor. We can then subtract the current limit we are trying to reach, and find the root of that function. We are going to receive multiple roots for this, so we select the smallest positive, real root.

## Conclusions
These changes will make the kinematics far more accurate, maximize our thrust envelope, and can account for parametric uncertainties - all while being relatively computationally inexpensive.

However, there will undoubtedly be changes and tweaks needed. Here is a short list of things which will be necessary to change in the future:
1. Current draw percentage cap: As stated in 7135693945228a3f3f5a52525e017ceebcf4fb64, we should likely limit how much current we draw to only be some percentage of the maximum current to account for any inaccuracies. The exact percentage will have to depend on how accurately we're able to model current as a function of thrust during testing.
2. Updated thrust to current data: Similar to the last step, we are going to need to run tests on our own setup to measure current draw more accurately.

There are certainly other issues which will arise, however the code in this branch works around the limitations that are known to the developer. Any changes that must be made likely won't be too difficult to fix (famous last words?). Please feel free to ask any questions you might have regarding the code or the methodology.

liamgilligan committed:
Reworked kinematics node to use numpy matrix multiplication

liamgilligan committed:
Removed commented-out code and added comments for new code

liamgilligan committed:
Updated Motor Configuration Matrix to use real values

See 61c1f87f00163e9e2e90469bd62be14f170dd806

liamgilligan committed:
Removed root3 variable as it is unused

liamgilligan committed:
Fixed Broken Matrix Multiplication

The list in the motor message (data) needed to be initialized to all zeroes, and each value from the result of the matrix multiplication must be assigned to that list. This is because numpy's matmul returns an ndarray and not a list, and therefore cannot be directly copied.

TODO: Use the  function on the resultant ndarray to do direct assignment.

liamgilligan committed:
Added TSV file for storing Thrust->Current Data

Data came from the Blue Robotics T200 Motor Technical Details, link will be in commit comment. Data will be updated with more accurate numbers after testing.

liamgilligan committed:
Added functions for converting Thrust->Current and generating parameters for previous function.

liamgilligan committed:
Add thrust limit scaling and fixed thrust to current data path

liamgilligan committed:
Set motor message data directly from numpy func call

liamgilligan committed:
Added thrust limit scaling and pilot thrust control

The thrust limit scalar determines maximum thrust in direction, maximum pilot input controls percentage of that maximum.

TODO: Implement current scaling and select the smaller of the scalars between current and thrust limits to determine thrust maximum

liamgilligan committed:
Added the ability to shift the center of mass by changing a ROS Parameter

When the center_of_mass_offset parameter is updated, the motor_config matrix will regenerate and a new inverse_config matrix will be generated from that.

liamgilligan committed:
Added functionality for respecting current draw limitations

Note: In production, we will have to gather our own thrust->current data, since the provided Blue Robotics data won't accurately reflect our setup. Also, we should not consider the hard limits to be our actual limit, and instead use some percentage of that maximum, such as 80% or 90%

liamgilligan committed:
Replaced lambdas and map/filter with list comprehensions

List comprehensions are generally faster and more readable, see Gudio blog post and diff, respectively.

Also removed unnecessary initilization of the motor_msg data array.

mike-matera commented:
When I try to run the nodes I get this:

```
[thrust-4] Traceback (most recent call last):
[thrust-4]   File "/home/maximus/Robotics/cabrillo_rov_2023/install/seahawk/lib/seahawk/thrust", line 33, in <module>
[thrust-4]     sys.exit(load_entry_point('seahawk', 'console_scripts', 'thrust')())
[thrust-4]   File "/home/maximus/Robotics/cabrillo_rov_2023/build/seahawk/seahawk_deck/thrust.py", line 288, in main
[thrust-4]     rclpy.spin(Thrust())
[thrust-4]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/__init__.py", line 222, in spin
[thrust-4]     executor.spin_once()
[thrust-4]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 739, in spin_once
[thrust-4]     self._spin_once_impl(timeout_sec)
[thrust-4]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 736, in _spin_once_impl
[thrust-4]     raise handler.exception()
[thrust-4]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/task.py", line 239, in __call__
[thrust-4]     self._handler.send(None)
[thrust-4]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 437, in handler
[thrust-4]     await call_coroutine(entity, arg)
[thrust-4]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 362, in _execute_subscription
[thrust-4]     await await_or_execute(sub.callback, msg)
[thrust-4]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 107, in await_or_execute
[thrust-4]     return callback(*args)
[thrust-4]   File "/home/maximus/Robotics/cabrillo_rov_2023/build/seahawk/seahawk_deck/thrust.py", line 275, in _callback
[thrust-4]     current_scalar = self.get_minimum_current_scalar(motor_msg.data)
[thrust-4]   File "/home/maximus/Robotics/cabrillo_rov_2023/build/seahawk/seahawk_deck/thrust.py", line 139, in get_minimum_current_scalar
[thrust-4]     total_scalar = self.get_current_scalar_value(mv, self.TOTAL_CURRENT_LIMIT)
[thrust-4]   File "/home/maximus/Robotics/cabrillo_rov_2023/build/seahawk/seahawk_deck/thrust.py", line 123, in get_current_scalar_value
[thrust-4]     return min(real_positive)
[thrust-4] ValueError: min() arg is an empty sequence
[ERROR] [thrust-4]: process has died [pid 91178, exit code 1, cmd '/home/maximus/Robotics/cabrillo_rov_2023/install/seahawk/lib/seahawk/thrust --ros-args -r __node:=thrust'].
```

liamgilligan committed:
Handle all-zero twist messages by publishing all-zero motor message

liamgilligan commented:
> When I try to run the nodes I get this:
> 
> ```
> [thrust-4] Traceback (most recent call last):
> [thrust-4]   File "/home/maximus/Robotics/cabrillo_rov_2023/install/seahawk/lib/seahawk/thrust", line 33, in <module>
> [thrust-4]     sys.exit(load_entry_point('seahawk', 'console_scripts', 'thrust')())
> [thrust-4]   File "/home/maximus/Robotics/cabrillo_rov_2023/build/seahawk/seahawk_deck/thrust.py", line 288, in main
> [thrust-4]     rclpy.spin(Thrust())
> [thrust-4]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/__init__.py", line 222, in spin
> [thrust-4]     executor.spin_once()
> [thrust-4]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 739, in spin_once
> [thrust-4]     self._spin_once_impl(timeout_sec)
> [thrust-4]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 736, in _spin_once_impl
> [thrust-4]     raise handler.exception()
> [thrust-4]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/task.py", line 239, in __call__
> [thrust-4]     self._handler.send(None)
> [thrust-4]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 437, in handler
> [thrust-4]     await call_coroutine(entity, arg)
> [thrust-4]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 362, in _execute_subscription
> [thrust-4]     await await_or_execute(sub.callback, msg)
> [thrust-4]   File "/opt/ros/humble/local/lib/python3.10/dist-packages/rclpy/executors.py", line 107, in await_or_execute
> [thrust-4]     return callback(*args)
> [thrust-4]   File "/home/maximus/Robotics/cabrillo_rov_2023/build/seahawk/seahawk_deck/thrust.py", line 275, in _callback
> [thrust-4]     current_scalar = self.get_minimum_current_scalar(motor_msg.data)
> [thrust-4]   File "/home/maximus/Robotics/cabrillo_rov_2023/build/seahawk/seahawk_deck/thrust.py", line 139, in get_minimum_current_scalar
> [thrust-4]     total_scalar = self.get_current_scalar_value(mv, self.TOTAL_CURRENT_LIMIT)
> [thrust-4]   File "/home/maximus/Robotics/cabrillo_rov_2023/build/seahawk/seahawk_deck/thrust.py", line 123, in get_current_scalar_value
> [thrust-4]     return min(real_positive)
> [thrust-4] ValueError: min() arg is an empty sequence
> [ERROR] [thrust-4]: process has died [pid 91178, exit code 1, cmd '/home/maximus/Robotics/cabrillo_rov_2023/install/seahawk/lib/seahawk/thrust --ros-args -r __node:=thrust'].
> ```

Fixed in be5a597e20436d111d9daaca00486783b43ada6d!

mike-matera committed:
Update kinematics debugging view.

mike-matera committed:
Updated motor ordering.

mike-matera committed:
Adding a bag for debugging.

mike-matera committed:
Fix bag recording

liamgilligan committed:
Changed Thrust->Current model to 6th degree polynomial and fixed 0 thrust test case

steph-lheureux commented on src/seahawk/seahawk_deck/thrust.py:
```diff
@@ -39,28 +44,223 @@ class Thrust(Node):
+        self.motor_config = self.generate_motor_config()
+        self.inverse_config = np.linalg.pinv(self.motor_config, rcond=1e-15, hermitian=False)
+
         self.subscription = self.create_subscription(Twist, 'drive/twist', self._callback, 10)
```
The `thrust` node should subscribe to topic `desired_twist` rather than `drive/twist`. `drive/twist` is an old name for this topic and has since been updated. See [`input_xbox_one.py`](https://github.com/CabrilloRoboticsClub/cabrillo_rov_2023/blob/02662b744abcd30fe495a150c504886342bb1732/src/seahawk/seahawk_deck/input_xbox_one.py#L92C58-L92C71) on main

steph-lheureux replied:
Also consider changing the quote type to double quotes to be consistent with the the rest of this file and project

steph-lheureux commented on src/seahawk/seahawk_deck/thrust.py:
```diff
@@ -39,28 +44,223 @@ class Thrust(Node):
+        self.inverse_config = np.linalg.pinv(self.motor_config, rcond=1e-15, hermitian=False)
+
         self.subscription = self.create_subscription(Twist, 'drive/twist', self._callback, 10)
         self.motor_pub = self.create_publisher(Float32MultiArray, 'drive/motors', 10)
```
Consider making the topic name more descriptive. And change string to use double quotes for consistency

steph-lheureux commented on src/seahawk/seahawk_deck/thrust.py:
```diff
@@ -39,28 +44,223 @@ class Thrust(Node):
+        self.__params = Thrust.__generate_curve_fit_params()
+
+    def get_polynomial_coef(self, mv: list, limit: float) -> list:
+        """
```
Add a short description of what the function does in the doc string

steph-lheureux commented on src/seahawk/seahawk_deck/thrust.py:
```diff
@@ -39,28 +44,223 @@ class Thrust(Node):
+                        else float('inf'))
+                    for thrust in motor_values])
 
     def _callback(self, twist_msg):
```
Double underscore for private methords

steph-lheureux commented on src/seahawk/seahawk_deck/thrust.py:
```diff
@@ -39,28 +44,223 @@ class Thrust(Node):
+                self.__params[5] * sum(mv),
+                self.__params[6] * len(mv) - limit]
+
+    def get_current_scalar_value(self, mv: list, limit: float) -> float:
```
Consider adding a link to the paper you wrote on the methodology. Add the pdf to the [doc](https://github.com/CabrilloRoboticsClub/cabrillo_rov_2023/tree/main/doc) directory

steph-lheureux replied:
Moved to Issue #194

steph-lheureux commented on src/seahawk/seahawk_deck/thrust.py:
```diff
@@ -39,28 +44,223 @@ class Thrust(Node):
+        esc2_scalar = self.get_current_scalar_value(mv[4:], self.ESC_CURRENT_LIMIT)
+
+        return min(total_scalar, esc1_scalar, esc2_scalar)
+
```
Too many new lines here

steph-lheureux commented on src/seahawk/seahawk_deck/thrust.py:
```diff
@@ -39,28 +44,223 @@ class Thrust(Node):
+            [-0.105, -0.15, -0.038]
+        ]
+
+        self.motor_thrusts = [
```
Consider adding a comment on this and `self.motor_positions` elaborating where the numbers are from/what they represent

steph-lheureux commented on src/seahawk/seahawk_deck/thrust.py:
```diff
@@ -39,28 +44,223 @@ class Thrust(Node):
+            self.motor_pub.publish(motor_msg)
+            return
+
         # +1 = Full thrust, Forwards
```
Delete this comment

liamgilligan committed:
Added descripton for motor position and thrust arrays

liamgilligan committed:
Updated twist subscription to desired_twist

liamgilligan committed:
Changed single quotes to double quotes for consistency

liamgilligan committed:
Changed publishing topic name to motor_values

liamgilligan committed:
Added function description for get_polynomial_coef

liamgilligan committed:
Fixed lack of and excess of newlines in certain areas

liamgilligan committed:
Deleted outdated motor config comment

liamgilligan committed:
Fixed bug for scaling based off twist and corrected spelling errors

steph-lheureux reviewed (approved):
The revisions made to the `Thrust` node in this PR are extremely impressive. These enhancements are a substantial improvement over the previous implementation. The updated methodology which uses a much more mathematically sound way of converting `Twist` to motor values will significantly improve the drive-ability of the robot. From reviewing this code, reading the paper you wrote, and talking to you over the last few months, it is clear that the underlying mathematics of the kinematics, current limiting, and thrust limiting code are well thought out and thoroughly researched. After viewing the simulation in rviz I am confident that this should be merged 

Also very noice list comprehensions :D

liamgilligan committed:
Cleaned up parameter callback function

Updates CoM offset member by calling self.get_parameter(...).value, and moves the other function calls within the try block

liamgilligan committed:
Removed CoM offset member variable

Removed as member was never accessed outside of parameter setting callback. Also replaced try/catch with a check for paramter validity and an early return.

mike-matera commented:
I'm getting this error:

```
[thrust-4] Traceback (most recent call last):
[thrust-4]   File "/home/maximus/Robotics/cabrillo_rov_2023/install/seahawk/lib/seahawk/thrust", line 33, in <module>
[thrust-4]     sys.exit(load_entry_point('seahawk', 'console_scripts', 'thrust')())
[thrust-4]   File "/home/maximus/Robotics/cabrillo_rov_2023/build/seahawk/seahawk_deck/thrust.py", line 278, in main
[thrust-4]     rclpy.spin(Thrust())
[thrust-4]   File "/home/maximus/Robotics/cabrillo_rov_2023/build/seahawk/seahawk_deck/thrust.py", line 80, in __init__
[thrust-4]     self.motor_config = self.generate_motor_config()
[thrust-4] TypeError: Thrust.generate_motor_config() missing 1 required positional argument: 'center_of_mass_offset'
```

liamgilligan committed:
Fixed initilization by passing in CoM offset param

liamgilligan committed:
Changed topic names in Input and RViz to match new topics in kinematics

liamgilligan committed:
Updated and reformatted motor position matrix

Updated according to Isaac's new values given. Also reformatted the motor thrust matrix to match the new formatting.

mike-matera commented:
This worked! Nice job.

liamgilligan committed:
Resolved merge conflicts
